### PR TITLE
Self load balancing (and complete reworking of the intra-cluster client)

### PR DESCRIPTION
**Pull request by dfellis**

### CODE REVIEW

**Review by aguillenv — APPROVED** [review]

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -60,6 +60,12 @@ jobs:
       with:
         node-version: 14.x
 
+    # The AVM requires openssl dev files to build now, and apparently the rust openssl code can auto-acquire it through vcpkg, so set that up
+    - name: Set up vcpkg
+      run: |
+        git clone https://github.com/microsoft/vcpkg
```
Comment (depombo) [inline .github/workflows/release.yml:66]: should we pin to a specific tag?

File: anycloud/cli/src/logger/mod.rs [hunk 2]
```diff
@@ -18,6 +18,7 @@ pub enum ErrorType {
   ScaleFailed = 120,
   PostFailed = 121,
   RunAgzFailed = 122,
+  DuplicateDnsPrivateIp = 123,
```
Comment (depombo) [inline anycloud/cli/src/logger/mod.rs:21]: we need to duplicate this in `deploy`. hmm maybe there is a way to get rid of this interop

File: avm/src/daemon/ctrl.rs [hunk 3]
```diff
@@ -169,25 +188,120 @@ impl ControlPort {
...
+      // TODO: Should this error propagate up to the stats loop or no?
+      return;
+    } else if self_vm_vec.len() > 1 {
+      // This hopefully never happens, but if it does, we need to change the daemon initialization
+      error!(
+        ErrorType::DuplicateDnsPrivateIp,
+        "Private IP address collision detected! I don't know who I really am!"
+      )
+      .await;
+      // TODO: Should this error propagate up to the stats loop or no?
```
Comment (depombo) [inline avm/src/daemon/ctrl.rs:260]: hmmm that is a good point

File: avm/src/vm/opcode.rs [hunk 4]
```diff
@@ -2982,16 +2988,107 @@ pub static OPCODES: Lazy<HashMap<i64, ByteOpcode>> = Lazy::new(|| {
   });
 
   async fn http_listener(req: Request<Body>) -> VMResult<Response<Body>> {
+    // Grab the headers
+    let headers = req.headers();
+    // Check if we should load balance this request
+    if !headers.contains_key("x-alan-rr") {
```
Comment (depombo) [inline avm/src/vm/opcode.rs:2994]: can we expand the comment a bit to explain how the load balancing works with this secret header?

**Review by depombo — APPROVED** [review]

File: avm/src/daemon/ctrl.rs [hunk 5]
```diff
@@ -169,25 +188,120 @@ impl ControlPort {
...
+      Some(self_vm) => self.lrh.get_leader_id() == self_vm.private_ip_addr,
+      None => false,
+    }
+  }
+
+  pub fn get_leader(self: &ControlPort) -> Option<&VMMetadata> {
+    match self.vms.get(self.lrh.get_leader_id()) {
+      Some(vm) => Some(&vm),
+      None => None,
+    }
```
Comment (cdmistman) [inline avm/src/daemon/ctrl.rs:304]: nit: could be
```suggestion
    self.vms.get(self.lrh.get_leader_id()).as_ref()
```

**Review by cdmistman — APPROVED** [review]

File: .github/workflows/release.yml (hunk 1 above)
Comment (dfellis) [reply .github/workflows/release.yml:66]: So this is a Microsoft repository for a Microsoft compiler and Microsoft operating system. [I am also following the official install instructions](https://docs.microsoft.com/en-us/cpp/build/install-vcpkg?view=msvc-160&tabs=windows) so I think this is fine?

File: anycloud/cli/src/logger/mod.rs (hunk 2 above)
Comment (dfellis) [reply anycloud/cli/src/logger/mod.rs:21]: `deploy` could have a build script to extract this definition from the Alan repo, or we could potentially have some sort of common definition (likely still in this repo, but easier than parsing Rust code and getting the relevant data).

File: avm/src/vm/opcode.rs (hunk 4 above)
Comment (dfellis) [reply avm/src/vm/opcode.rs:2994]: Sure

File: avm/src/daemon/ctrl.rs (hunk 5 above)
Comment (dfellis) [reply avm/src/daemon/ctrl.rs:304]: Wouldn't that be `&Option<VMMetadata>` instead of `Option<&VMMetadata>` though?

Comment (cdmistman) [reply avm/src/daemon/ctrl.rs:304]: no - `as_ref` inverts the reference so it turns a `&Option<VMMetadata>` into a `Option<&VMMetadata>`

File: anycloud/cli/src/logger/mod.rs (hunk 2 above)
Comment (depombo) [reply anycloud/cli/src/logger/mod.rs:21]: Yeah, probably just a JSON file here

Comment (dfellis) [reply anycloud/cli/src/logger/mod.rs:21]: > Yeah, probably just a JSON file here

The issue is turning that JSON file into an enum in Rust will also be difficult.

**Pull request merged**